### ISSUE TITLE
Thinking toggle is a silent no-op on models whose chat template ignores enable_thinking (LFM2.5)

### ISSUE THREAD

**Issue by Helldez** [Helldez/BigMoeOnEdge#82]
## Summary

Turning Thinking off in the app (`--no-think` on the CLI) has no effect on LFM2.5-8B-A1B: the
model reasons anyway, and nothing tells the user the setting was ignored.

## Root cause

`--no-think` does exactly one thing — `inputs.enable_thinking = false`
(`core/src/engine/session.cpp:635`) — which llama.cpp forwards to the model's own Jinja chat
template as a template variable (`common/chat.cpp:895`). Whether that suppresses reasoning is
entirely up to the template.

LFM2.5's `chat_template.jinja` never references `enable_thinking`. Its only related variable is
`preserve_thinking`, which controls whether *past* turns' reasoning is re-rendered, not whether the
model reasons on this turn. So the rendered prompt is byte-identical with thinking on and off, and
the model — a reasoning model — reasons.

## Scope: three regimes, not two

| Family | Template honors `enable_thinking` | Outcome |
|---|---|---|
| Qwen3 / Qwen3.6 | yes | toggle works |
| gpt-oss (harmony) | no | works anyway — `session.cpp:651` special-cases it by priming the final channel |
| LFM2 / LFM2.5 | no | **toggle is a silent no-op** |

Gemma 4 does not raise the question (not a reasoning model). The failure is not specific to LFM2.5:
any architecture whose template ignores the variable lands here, and always silently.

## Note on the cost

The reasoning is still parsed out into the collapsible thinking block — the parser gate keys on the
presence of `<think>` in the template source, not on `enable_thinking` (`common/chat.cpp:1708`). So
the answer is not polluted. The real cost is that the model spends its token budget reasoning when
the user asked it not to.

## Trap for whoever fixes this

llama.cpp appears to already offer the detector: `common_chat_templates_support_enable_thinking()`
(`common/chat.cpp:355`). **It does not answer this question.** It returns `params.supports_thinking`,
which the LFM2 handler hardcodes to `true` (`common/chat.cpp:1696`) — it reports "this model can
reason", not "this template honors the flag".

## Direction (not yet decided)

1. Detect generically: render the prompt with `enable_thinking` true and false and compare. Identical
   renders mean the flag has no purchase on this model — surface that in the UI instead of offering a
   toggle that does nothing.
2. Prefill, as already done for harmony: close the reasoning span in the generation prompt
   (`<think></think>`) so the model proceeds to the answer. Must key on the template, not the model
   name (see CLAUDE.md: no model-specific constants).

Prior art in llama.cpp's own server and in other on-device apps is being surveyed before choosing.

## Repro

LFM2.5-8B-A1B-Q5_K_M.gguf, app Thinking toggle off, any prompt. Observed on 0.11.1-dev.

**Comment by Helldez** [comment]
## The mechanism already exists upstream — and is already in the pinned submodule

Neither of the two directions proposed above is what llama.cpp actually does. The upstream answer is
better than both, needs no per-model knowledge, and does not depend on the template honoring
anything. It splits cleanly into detection and enforcement.

**Detection is data, supplied by the model.** When a chat template is applied, the per-template
handler fills `common_chat_params.thinking_start_tag` / `thinking_end_tag`. The LFM2/LFM2.5 handler
sets them to `<think>` / `</think>` (`common/chat.cpp:1703-1704`). llama-server's entire test for
"can thinking be controlled on this model" is:

```cpp
// tools/server/server-common.cpp:1124
if (!chat_params.thinking_end_tag.empty()) {
    llama_params["reasoning_budget_tokens"]    = reasoning_budget;
    llama_params["reasoning_budget_start_tag"] = chat_params.thinking_start_tag;
    llama_params["reasoning_budget_end_tag"]   = chat_params.thinking_end_tag;
}
```

No model names, no template string matching on our side: the loaded model answers the question.

**Enforcement is a sampler, not a prompt trick.** `common_reasoning_budget_init()`
(`common/reasoning-budget.h:34`) returns a `llama_sampler *` running a state machine
IDLE → COUNTING → WAITING_UTF8 → FORCING → DONE. It watches for the start tag; with budget 0 it
transitions straight to FORCING (`common/reasoning-budget.cpp:69-73`) and emits the end-tag tokens
with every other logit driven to -inf. Budget semantics: `-1` unlimited, `0` close immediately,
`N` allow N tokens then force the close, optionally injecting `reasoning_budget_message` first.

Because it acts on logits, it never asks the template's permission — which is precisely the failure
in this issue. `common/reasoning-budget.cpp` is already in the pinned submodule's build
(`common/CMakeLists.txt:97`).

## Two wrinkles specific to this engine

1. **The sampler chain is null by default.** `Session::open()` builds a chain only when `temp > 0`;
   `temp <= 0` leaves `im.smpl == nullptr` and the decode loop on argmax (`session.cpp:381-391`,
   `session.cpp:840`) — deliberately, so the byte-identity gates stay deterministic. A reasoning
   budget must apply in argmax mode too, so it cannot simply be appended to the optional chain.
2. **Prefill tokens must be fed to the sampler.** The IDLE → COUNTING transition needs to observe the
   start tag. When the template opens `<think>` in the generation prompt, the tag is never sampled,
   so upstream replays prefill tokens into the sampler (`common/sampling.cpp:307-310`). Skipping this
   silently disables the budget on exactly the models that need it most.

Note `common_reasoning_budget_init` lives in `common/` (the non-stable layer), but it returns a plain
`llama_sampler *` built on the public sampler interface, and this engine already depends on `common/`
for the chat templates and reasoning parser. This is consistent with existing practice, not a new
coupling — and it is not a llama.cpp patch, so hard rule 1 is untouched.

**Comment by Helldez** [comment]
## Survey of prior art — and a correction to the direction above

Findings below were read from the pinned submodule and from the models' own template files, not from
documentation. Two of them change the plan.

### Correction: render-twice-and-diff is not a workaround, it is what llama.cpp does

Direction 1 was dismissed too quickly above. llama.cpp implements exactly it, in
`common/chat-diff-analyzer.cpp` (`analyze_reasoning::compare_thinking_enabled()`): it renders the
template with `enable_thinking` false, then again with true, and diffs. It handles both polarities
(marker added when true, marker added when false) and falls back to a tail-anchored comparison when
the diff is noisy — the comment names the real case, SmolLM3, whose *system message* changes when the
flag flips. It is documented as rule R2 in `docs/autoparser.md`. This is the path for templates it
does not recognize.

So detection and enforcement are two questions, not competing answers to one:

- *can this model be silenced?* → the diff probe
- *how do I actually silence it?* → the budget sampler

Both are needed. The sampler alone leaves the UI unable to say when the toggle is real; the probe
alone still cannot stop a model that has no off switch.

### The upstream detector is worse than described above

The earlier warning about `common_chat_templates_support_enable_thinking()` understated the problem.
For any template matching a known handler, `supports_thinking` is not measured at all — it is a
hardcoded literal, `true` at nine sites (`chat.cpp:1023, 1149, 1293, 1544, 1696, 1865, 2078, 2410`)
and `false` at one. The empirical diff runs *only* for unrecognized templates (`chat.cpp:2772`).

LFM2 matches a known handler, so the constant applies: upstream would report "thinking is supported"
for LFM2.5 and be wrong about the only thing we care about. Of the 59 bundled templates, 20 reference
`enable_thinking`; for the other 39 the value is injected into the render context and silently
discarded.

### Upstream will not fix this

ggml-org/llama.cpp#20196 is the same failure on MiniMax-M2.5 — `--reasoning-budget 0` and the kwarg
both accepted, server logged `thinking = 0`, model reasoned anyway. **Closed as not planned.** Related
and also closed: #20409 (Qwen3.5), #21338 (Gemma-4). #23351 (per-request reasoning toggle) is open
with no maintainer reply. This is the steady state, not a pending patch to wait for.

### Confirmation on our model

Re-fetched `LiquidAI/LFM2.5-8B-A1B/raw/main/chat_template.jinja` directly. It reads
`preserve_thinking` (line 2, defaulted false), `messages`, `tools`, `add_generation_prompt`. It does
**not** reference `enable_thinking`. The `preserve_thinking` branches (lines 57, 64) gate whether
*prior* turns' reasoning is re-rendered, keyed on `loop.index0 > ns.last_user_index` — history, not
the current turn. The original diagnosis holds.

Worth noting `preserve_thinking` is itself a llama.cpp-injected variable, set by
`caps_apply_preserve_reasoning()` (`common/jinja/caps.cpp:22-26`) for `--reasoning-preserve`. LFM2.5's
template cooperates with llama.cpp's history-preservation feature but implements no thinking toggle.

### What on-device apps do

- **PocketPal AI** (`src/utils/thinkingCapabilityDetection.ts`) detects dynamically with no model
  table: renders once with `enable_thinking: true` and checks whether a `thinking_start_tag` comes
  back. Its precedence is `user > learned > detected`, where *learned* is revised from observed
  output, and it **fails open** — unsure means thinking stays on. It also ships a stripper for when
  the model reasons regardless. That last piece is an admission that no static detector is reliable.
- **Ollama** (`server/images.go:215`) substring-sniffs the template for `<think>`/`</think>` plus a
  hardcoded `gpt-oss` rule. Note the asymmetry: requesting thinking on an unsupported model is a hard
  400, while requesting it *off* is always silently accepted — the same blind spot as this issue.
- **ChatterUI** does no detection; the checkbox is forwarded blind.
- **LM Studio** relies on an author-declared field in `model.yaml`, unverified at runtime.

### Revised plan

1. **Budget sampler** — the enforcement, and the only part a template cannot ignore.
2. **Diff probe** — so the UI can distinguish a real toggle from a decorative one. Fail open.
3. *(later, optional)* A genuine upstream contribution is available and small. `common/jinja/caps.cpp`
   instruments the Jinja interpreter to record whether a variable was actually *read* during a render
   (`context::is_get_stats`), which answers this question exactly rather than by inference. There is
   no `supports_enable_thinking` entry in `caps::to_map()`, though the sibling
   `supports_preserve_reasoning` sits beside it and `chat_template_caps` is already exposed over
   `/props`. Worth proposing only after our own implementation has proven itself.

**Comment by Helldez** [comment]
## On-device measurement: forcing the close is not sufficient on LFM2.5

The attempt in #83 is closed. The reason is worth recording, because it rules out the approach that
looked most promising in the comments above — and it was only visible on a real model, not in the
host gates.

### The plan was: budget-0 reasoning sampler

`common_reasoning_budget_init` with budget 0 closes the reasoning block on logits the instant the
model opens it, needing no cooperation from the template. That part works exactly as advertised: the
block is closed, immediately, every time.

### What the model actually does

LFM2.5-8B-A1B-Q5_K_M, `--chatml --no-think`, empty forced message (the upstream default), asking a
question that needs no reasoning at all:

```
<think></think>
<think></think>
Okay, the user asked, "What is 2+2?" and wants a brief answer. Let me think. Well, in standard
arithmetic, 2 plus 2 equals 4. [... ~130 more tokens of reasoning ...] Alright, the answer is 4.
</think>
2+2 = 4.
```

Read that carefully:

1. The model opens the block; the sampler closes it. Empty block.
2. The model **re-opens** it; the sampler closes it again. Empty block. (The sampler re-arms on a new
   start tag by design — `reasoning-budget.cpp:124-139`.)
3. Denied twice, the model **stops using the tags** and reasons in plain prose.
4. It emits a lone `</think>` at the end, still believing it was inside a block.
5. Only then does it answer.

166 tokens generated, the reasoning intact, now sitting in the **answer** instead of a thinking
block, followed by a stray marker.

**This is worse than the bug it was meant to fix.** The reasoning was never suppressed — it was only
stripped of the tags that made it separable. Before the change, the reasoning at least parsed into a
collapsible block; after it, it is answer text.

### Why the empty-block detail matters

An empty `<think></think>` is not a thought a model recognizes as its own. Injecting a short
concluding line before the forced close instead of closing on nothing changed the behaviour
completely — the model answered directly, 16 tokens instead of 166. That is a real signal about
*why* the mechanism fails, and it points at what the model is reacting to.

It is not, however, a fix I am willing to ship as it stood: the line was English, invented, tuned by
trial and error against a single model, and hardcoded into the engine. Upstream's own
`--reasoning-budget-message` defaults to empty precisely because that wording is a caller's decision.
If a message is part of the answer, it has to be configuration, not a constant.

### What this rules out

- **Budget 0 alone is not an answer for this family.** Correct at the logit level, ineffective at the
  behavioural level. Any future attempt should be validated against generated *content*, not against
  "was the block closed".
- **Post-hoc trimming of the output is not an answer either.** The parser's reasoning rule is a
  single anchored `optional(...)` (`chat.cpp` lfm2 handler ~1724-1738), so multiple blocks, or any
  whitespace before the first one, leave literal markers in the content. Cutting those out by hand in
  the engine is fragile and belongs nowhere near a submodule bump.

### Still open

- Whether a caller-supplied budget message is the honest form of this (configuration, with the engine
  choosing nothing).
- Whether the right outcome for a model that refuses to be silenced is to **report it as
  uncontrollable** and say so in the UI, rather than fight it. The detection half of #83 was built for
  exactly that and is the part that still looks sound — though it may be over-built: llama-server runs
  no probe at all, it just checks whether a thinking end tag exists.
- Whether suppression belongs at the prompt level for these templates (pre-closing a block in the
  generation prompt) rather than at the logit level.

Measurements are reproducible with `bmoe-cli --chatml --no-think --progress -n 200` against the model
above; the closed branch `feat/think-control-82` carries the code they were taken with.

**Comment by Helldez** [comment]
## Shipped in v0.13.0 (#85) — leaving this open deliberately

The silence is fixed; the LFM2.5 toggle is not, and cannot be.

**What ships.** The engine renders the model's template at load and reports what that model will
actually honour, as `think_ctl` on `BMOE_READY`: `template` (the template reads the flag),
`prefill` (it does not, but reasoning is a structural section the turn can start past), or `none`
(the model owns its reasoning span and cannot be asked to skip it). The app disables the Thinking
switch, with the reason, when it is `none`. Nothing is keyed on model names — the `prefill`/`none`
split is read off `common_chat_params::thinking_start_tag`/`thinking_end_tag`, supplied by the model.

Where the turn does start past the reasoning, llama.cpp's own handler writes the span, so the engine
names no family's markers. That also retired the two hardcoded harmony strings this issue's "Trap"
section warned about — gpt-oss now takes the generic path and behaves identically.

**Why LFM2.5 is `none` rather than fixed.** Both directions proposed above were tried and measured:

- Direction 2 (prefill a closed span) *was* implemented and does render correctly — and LFM2.5
  reasons straight past it, emitting the reasoning **untagged into the answer**. Strictly worse than
  this bug, where it at least parsed into a collapsible block. Same end state as the budget-0
  attempt in #83, reached from the prompt side instead of the logit side.
- Direction 1 (render-and-diff) shipped, and is what makes the honest answer possible.

Ruled out along the way: this is not llama.cpp's LFM2.5 template-detection bug (#23852). Our pin
carries that fix, detection succeeds on this gguf, the template inside the model file is byte-identical
to the vendored one, and the rendered span had the LFM2 handler's own shape. Upstream's position on
suppression generally is ggml-org/llama.cpp#20196, closed *not planned*, and Liquid ships a separate
non-reasoning checkpoint rather than an off switch. LFM2.5 always reasoning is the model's design.

**What would close this** is a way to make LFM2.5 answer directly without the engine inventing text.
#83 measured that a short concluding line does it (16 tokens vs 166), but that line has to come from
the caller, not from a constant in the engine. Until that exists as configuration, `none` is the
truthful answer and the UI says it out loud.
